kafka-streams-framework/src/main/java/org/hypertrace/core/kafkastreams/framework/AsyncTransformer.java

```diff
@@ -0,0 +1,92 @@
+package org.hypertrace.core.kafkastreams.framework;
+
+import com.google.common.util.concurrent.RateLimiter;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import java.time.Duration;
+import java.util.List;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadFactory;
+import lombok.SneakyThrows;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.kstream.Transformer;
+import org.apache.kafka.streams.processor.ProcessorContext;
+
+@Slf4j
+@SuppressWarnings("UnstableApiUsage")
+public abstract class AsyncTransformer<K, V, KOUT, VOUT>
+    implements Transformer<K, V, KeyValue<KOUT, VOUT>> {
+  private final Executor executor;
+  private final BlockingQueue<CompletableFuture<List<KeyValue<KOUT, VOUT>>>> pendingFutures;
+  private final RateLimiter rateLimiter;
+  private ProcessorContext context;
+
+  // TODO: configurable executor - supplier pattern. This enables to use a common thread-pool for
+  // all stream tasks
+  public AsyncTransformer(int concurrency, int maxBatchSize, Duration flushInterval) {
+    ThreadFactory threadFactory =
+        new ThreadFactoryBuilder()
+            .setNameFormat("async-transformer-pool-%d")
+            .setDaemon(true)
+            .build();
+    this.executor = Executors.newFixedThreadPool(concurrency, threadFactory);
+    this.pendingFutures = new ArrayBlockingQueue<>(maxBatchSize);
+    this.rateLimiter = RateLimiter.create(1.0 / flushInterval.toSeconds());
+  }
+
+  @Override
+  public void init(ProcessorContext context) {
+    this.context = context;
+  }
+
+  public abstract List<KeyValue<KOUT, VOUT>> asyncTransform(K key, V value);
+
+  @SneakyThrows
+  @Override
+  public KeyValue<KOUT, VOUT> transform(K key, V value) {
+    // Flush based on time duration
+    if (rateLimiter.tryAcquire()) {
+      log.info("flush start - type: time, queue size: {}", pendingFutures.size());
+      processResults();
+      log.info("flush end - type: time, queue size: {}", pendingFutures.size());
+    }
+
+    CompletableFuture<List<KeyValue<KOUT, VOUT>>> future =
+        CompletableFuture.supplyAsync(() -> asyncTransform(key, value), executor);
+    // thread blocked when queue is full. queue consumer runs in this same thread.
```

Comment:
Why do we want the queue consumer running in the same thread, doesn't that defeat the purpose of making this async? We're basically saying keep accepting things and then block on some interval until we're all caught up, all we're doing is making things step-wise. It would be better to hold messages and check them periodically for completion on some other thread. Is there a requirement to keep messages in-order (if so, I'm not sure I understand the purpose of making this async in the first place)

Reply:
Entire kafka streams framework is built with data structures (context/stores/etc) which are not thread safe.
Every thing happens in stream thread. Including punctuator. Its prohibited to use any of these in other threads.
The moment we access context in a different thread, most of the things go wrong.

The intention here is to make some application code async which is taking more time to process asynchronously (grpc calls in our case). 

I added a unit test to demonstrate the gains we get by this. If possible pls take a look and run the test. If needed, I can give a demo.

Reply:
So this is the fundamental question, will stick to this thread.


> I added a unit test to demonstrate the gains we get by this. If possible pls take a look and run the test. If needed, I can give a demo.

I did take a look at this. I get that if the transform function is holding the thread for a long time splitting out to multiple will help performance (although a sleep isn't a great example of that, since in practice we're typically cpu bound even on a small number of threads). I guess a better way of phrasing what I was getting at is that we're only solving half the problem, and we're unnecessarily introducing latency.

If we want to stay with one main thread interacting with kafka, that's fine. That main thread should be the one queueing up work, and dumping that work back to kafka. That doesn't really force a batch processor though unless there's a major reason like reducing commit frequency that I'm unaware of. As it is, I would thing the batch processor will unnecessarily delay output (a transform result isn't output when it's ready, it's output once both the batch is full [time or size] and every result in front of it is output) and thus we'll get into a state where we're holding whole batches in memory creating unnecessary pressure. This could easily be made more real-time without batches. Imagine an accept workflow that at a high level works like:

- attempt to schedule process of incoming message. A bounded blocking queue would be a good fit here, as it'll automatically wait once full (via `.put` or `.offer` w/ timeout) so the app temporarily stops accepting new messages
- after scheduling is successful, drain a second result queue (which holds results directly, outputting each, committing if nonzero.

The process message job task would call the main app logic, then once a result is returned, add it to the result queue and remove from process queue. An alternative implementation could be to use a single schedule queue (no separate result) and just check it via iteration.

Reply:
Discussed offline - the big misunderstanding here is that commit cannot be called with any queue since it looks at the last accepted message, not the last output one. Given that, we need an empty queue when we commit, which basically means buffer and drain. We could eagerly forward results if we really want to as long as we wait to commit until we drain, but not sure the advantage of that. Seems nicer to move commit closer to the actual response send.